File: web/src/lib/api/models/app/oauth/OAuthApp.ts
// OAuth 2.1 authorization-server types (mirror of internal/models/oauth_app.go).

export type OAuthAppStatus = "active" | "disabled";

export interface OAuthApplication {
    id: string;
    organization_id: string;
    created_by: string;
    name: string;
    description: string;
    logo_url: string;
    website_url: string;
    client_id: string;
    redirect_uris: string[];
    // Domains this app's registered webhooks must point at. A leading dot
    // (.acme.com) matches subdomains; a bare domain is an exact match. Empty
    // forbids the app from registering webhooks.
    allowed_webhook_domains: string[];
    // The app's webhook callback URL. Must be HTTPS and its host must be inside
    // allowed_webhook_domains; empty means the app receives no events.
    webhook_url: string;
    // Event types the app subscribes to (empty = all events the granting org allows).
    webhook_events: string[];
    // Bitmask of the API permissions this app may request (same bits as API keys).
    scopes: number;
    status: OAuthAppStatus;
    created_at: string;
    updated_at: string;
}

// Returned once on create / secret rotation; client_secret is shown a single time.
export interface OAuthApplicationWithSecret extends OAuthApplication {
    client_secret?: string;
}

export interface OAuthApplicationsResult {
    applications: OAuthApplication[];
}

export interface OAuthApplicationInput {
    name: string;
    description?: string;
    logo_url?: string;
    website_url?: string;
    redirect_uris: string[];
    allowed_webhook_domains?: string[];
    webhook_url?: string;
    webhook_events?: string[];
    scopes: number;
}

// The consent screen payload (GET /oauth/authorize/details).
export interface OAuthConsentInfo {
    client_id: string;
    name: string;
    description: string;
    logo_url: string;
    website_url: string;
    redirect_uri: string;
    scopes: string[];
    state: string;
}

// An app the current user has authorized (GET /oauth/authorized-apps).
export interface OAuthAuthorizedApp {
    application_id: string;
    name: string;
    logo_url: string;
    website_url: string;
    scopes: number;
    authorized_at: string;
    last_used_at?: string;
}

export interface OAuthAuthorizedAppsResult {
    authorized_apps: OAuthAuthorizedApp[];
}

File: internal/models/oauth_app.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// OAuth2 authorization server domain types. Apps register as OAuth clients;
// users grant them scoped access via the authorization-code flow (client secret
// required, PKCE optional); the issued access token carries an API-permission
// bitmask (Scopes) and authenticates API calls through the same gates as an API key.

type OAuthAppStatus string

const (
	OAuthAppActive   OAuthAppStatus = "active"
	OAuthAppDisabled OAuthAppStatus = "disabled"
)

// Credential prefixes mirror the api_keys `wmbly_` convention so a leaked token
// is greppable and self-describing.
const (
	OAuthClientIDPrefix     = "wmcid_"
	OAuthClientSecretPrefix = "wmcs_"
	OAuthAccessTokenPrefix  = "wmat_"
	OAuthRefreshTokenPrefix = "wmrt_"
	OAuthCodePrefix         = "wmac_"
)

// Lifetimes. The authorization code is single-use and short; access tokens are
// short-lived; refresh tokens are long-lived and rotate on every exchange.
const (
	OAuthAuthorizationCodeTTL = 10 * time.Minute
	OAuthAccessTokenTTL       = time.Hour
	OAuthRefreshTokenTTL      = 90 * 24 * time.Hour
)

// OAuthApplication is a registered third-party OAuth client.
type OAuthApplication struct {
	ID               uuid.UUID      `json:"id"`
	OrganizationID   uuid.UUID      `json:"organization_id"`
	CreatedBy        uuid.UUID      `json:"created_by"`
	Name             string         `json:"name"`
	Description      string         `json:"description"`
	LogoURL          string         `json:"logo_url"`
	WebsiteURL       string         `json:"website_url"`
	ClientID         string         `json:"client_id"`
	ClientSecretHash string         `json:"-"`
	RedirectURIs     []string       `json:"redirect_uris"`
	Scopes           uint64         `json:"scopes"`
	Status           OAuthAppStatus `json:"status"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
}

// OAuthApplicationWithSecret is returned exactly once, on create or secret
// rotation; the plaintext secret is never stored or shown again.
type OAuthApplicationWithSecret struct {
	OAuthApplication
	ClientSecret string `json:"client_secret,omitempty"`
}

// OAuthApplicationWrite is the create/update payload from the developer UI.
type OAuthApplicationWrite struct {
	Name         string   `json:"name"`
	Description  string   `json:"description"`
	LogoURL      string   `json:"logo_url"`
	WebsiteURL   string   `json:"website_url"`
	RedirectURIs []string `json:"redirect_uris"`
	Scopes       uint64   `json:"scopes"`
}

// OAuthAuthorizationCode is a single-use code bound to a PKCE challenge and the
// exact scopes/redirect the user consented to.
type OAuthAuthorizationCode struct {
	ID                  uuid.UUID
	CodeHash            string
	ApplicationID       uuid.UUID
	OrganizationID      uuid.UUID
	UserID              uuid.UUID
	RedirectURI         string
	Scopes              uint64
	CodeChallenge       string
	CodeChallengeMethod string
	UsedAt              *time.Time
	ExpiresAt           time.Time
	CreatedAt           time.Time
}

// OAuthAccessGrant is an issued access+refresh token pair (tokens stored hashed).
type OAuthAccessGrant struct {
	ID               uuid.UUID  `json:"id"`
	ApplicationID    uuid.UUID  `json:"application_id"`
	OrganizationID   uuid.UUID  `json:"organization_id"`
	UserID           uuid.UUID  `json:"user_id"`
	Scopes           uint64     `json:"scopes"`
	AccessTokenHash  string     `json:"-"`
	RefreshTokenHash string     `json:"-"`
	AccessExpiresAt  time.Time  `json:"access_expires_at"`
	RefreshExpiresAt *time.Time `json:"refresh_expires_at,omitempty"`
	RevokedAt        *time.Time `json:"revoked_at,omitempty"`
	LastUsedAt       *time.Time `json:"last_used_at,omitempty"`
	CreatedAt        time.Time  `json:"created_at"`
}

// OAuthAuthorizedApp is one row in a user's "apps you've authorized" list (a
// grant joined to its application's display fields).
type OAuthAuthorizedApp struct {
	ApplicationID uuid.UUID  `json:"application_id"`
	Name          string     `json:"name"`
	LogoURL       string     `json:"logo_url"`
	WebsiteURL    string     `json:"website_url"`
	Scopes        uint64     `json:"scopes"`
	AuthorizedAt  time.Time  `json:"authorized_at"`
	LastUsedAt    *time.Time `json:"last_used_at,omitempty"`
}
